rclcpp/src/rclcpp/intra_process_manager.cpp

```diff
@@ -225,5 +225,55 @@ IntraProcessManager::can_communicate(
   return true;
 }
 
+size_t
+IntraProcessManager::lowest_available_capacity(const uint64_t intra_process_publisher_id) const
+{
+  size_t capacity = std::numeric_limits<size_t>::max();
+
+  auto publisher_it = pub_to_subs_.find(intra_process_publisher_id);
+  if (publisher_it == pub_to_subs_.end()) {
+    // Publisher is either invalid or no longer exists.
+    RCLCPP_WARN(
+      rclcpp::get_logger("rclcpp"),
+      "Calling lowest_available_capacity for invalid or no longer existing publisher id");
+    return 0u;
+  }
+
+  if (publisher_it->second.take_shared_subscriptions.empty() &&
+    publisher_it->second.take_ownership_subscriptions.empty())
+  {
+    // no subscriptions available
+    return 0u;
+  }
+
+  for (const auto sub_id : publisher_it->second.take_shared_subscriptions) {
+    capacity = std::min(capacity, available_capacity(sub_id));
```

Comment:
I think that after the change `max -> 0`, this is now broken.
the current implementation of `available_capacity` will return 0 if a subscription is not found, but this is not correct here.

Let's say that a publisher has 2 subscriptions; if one goes out of scope and the other has capacity 2, the API should return 2.

I recommend to also add unit-tests for these corner cases

Reply:
that's a good hint. I've made changes to handle subscriptions which went out of scope and added an unit test.